### PR TITLE
dashboard callout

### PR DESCRIPTION
**Pull request by simleap**
screenshot: https://goo.gl/aVK5gi

### CODE REVIEW

File: app/assets/javascripts/dashboard/base.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      parent = $(this).closest('.js-metadata-fields')
+      parent.find('input').val('');
+      parent.removeClass('d-flex');
+      parent.hide();
+    })
+  }
+
+  this.init = function() {
+    onCLickJSAddMetadataField();
+    onClickJSRemoveMetadataField();
```
Comment (samnang) [inline app/assets/javascripts/dashboard/base.js]: Should we better to have a class to handle metadata field base on data behavior attribute.

new MetaDataFields($("[data-behavior='metadata-fields']"))

File: app/assets/javascripts/dashboard/callout.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+  dashboardBase = new DashboardBase();
+
+  this.init = function() {
+    dashboardBase.init();
+  }
+}
+
+$(document).on('turbolinks:load', function() {
+  if (!(page.controller() === 'callouts')) {
```
Comment (samnang) [inline app/assets/javascripts/dashboard/callout.js]: can we use `!==`?

File: app/controllers/dashboard/access_tokens_controller.rb [hunk 3]
```diff
@@ -1,5 +1,5 @@
 class Dashboard::AccessTokensController < Dashboard::BaseController
   def index
-    @access_tokens = current_account.access_tokens
+    @access_tokens = current_account.access_tokens.page(params[:page]).per(10)
```
Comment (samnang) [inline app/controllers/dashboard/access_tokens_controller.rb]: Do we actually have so many access tokens to have pagination on this page?

File: app/assets/stylesheets/dashboard/callouts.scss [hunk 4]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Dashboard::Callouts controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (samnang) [inline app/assets/stylesheets/dashboard/callouts.scss]: if we don't have any codes, we should delete this file.

File: app/controllers/dashboard/contacts_controller.rb [hunk 5]
```diff
@@ -2,7 +2,7 @@ class Dashboard::ContactsController < Dashboard::BaseController
   before_action :set_contact, only: [:show, :edit, :update, :destroy]
 
   def index
-    @contacts = current_account.contacts
+    @contacts = current_account.contacts.page(params[:page]).per(10)
```
Comment (samnang) [inline app/controllers/dashboard/contacts_controller.rb]: We should have consistency in global configuration across models.

https://github.com/kaminari/kaminari#configuring-kaminari

Seem 25 by default is good, but if you want to customize into 20 is fine with me, but put it in global config.

File: app/controllers/dashboard/users_controller.rb [hunk 6]
```diff
@@ -1,5 +1,5 @@
 class Dashboard::UsersController < Dashboard::BaseController
   def index
-    @user = current_account.users
+    @users = current_account.users.page(params[:page]).per(10)
```
Comment (samnang) [inline app/controllers/dashboard/users_controller.rb]: No need to specify the `per` option here.

File: app/views/dashboard/callouts/edit.html.haml [hunk 7]
```diff
@@ -0,0 +1,3 @@
+.col-12.col-md-8
+  %h5.card-title Edit callout
+  = render partial: 'form', locals: { callout: @callout, url: dashboard_callout_path(@callout) }
```
Comment (samnang) [inline app/views/dashboard/callouts/edit.html.haml]: I don't think you need to pass url because Rails will generate a correct base on state of model

you should just password `simple_form_for [:dashboard, callout]`

File: app/views/dashboard/callouts/show.html.haml [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        %dd.col-sm-7= @callout.id
+      %dl.row
+        %dt.col-sm-5.text-md-right Status:
+        %dd.col-sm-7= @callout.status
+      %dl.row
+        %dt.col-sm-5.text-md-right Meta data:
+        %dd.col-sm-7= infinit_list(@callout.metadata)
+      %dl.row
+        %dt.col-sm-5.text-md-right Created at:
+        %dd.col-sm-7= @callout.created_at
```
Comment (samnang) [inline app/views/dashboard/callouts/show.html.haml]: I feel it doesn't layout nicely

localhost_3000_dashboard_callouts_1

File: app/views/kaminari/_gap.html.haml [hunk 9]
```diff
@@ -0,0 +1,2 @@
+%li.page-item.disabled
+  = link_to raw(t 'views.pagination.truncate'), '#', class: 'page-link'
```
Comment (samnang) [inline app/views/kaminari/_gap.html.haml]: Do we actually need to customize kaminari view template?

File: spec/system/devises/invitations/new_spec.rb [hunk 10]
```diff
@@ -1,16 +1,16 @@
...
   end
 
   it 'will send an invitation (pending)' do
-    # visit '/users/invitation/new'
-    # fill_in 'user[email]', with: 'bopha@somleng.com'
-    # click_button 'Send an invitation'
-    #
-    # expect(page).to have_text('sent an invitation successfully.')
+    visit '/users/invitation/new'
+    fill_in 'user[email]', with: 'bopha@somleng.com'
```
Comment (samnang) [inline spec/system/devises/invitations/new_spec.rb]: Better use label instead of directly with internal field.

File: spec/system/devises/invitations/edit_spec.rb [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+  private
+
+  def raw_token
+    User.invite!(
+      { email: 'bophasd@somleng.com', account_id: inviter.account_id },
+      inviter
+    ).raw_invitation_token
+  end
+end
```
Comment (samnang) [inline spec/system/devises/invitations/edit_spec.rb]: This spec and new invitation should be merged.

File: spec/system/devises/invitations/edit_spec.rb [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    click_button 'Save'
+
+    expect(page).to have_text('Your password was set successfully. You are now signed in.')
+  end
+
+  private
+
+  def raw_token
+    User.invite!(
+      { email: 'bophasd@somleng.com', account_id: inviter.account_id },
```
Comment (samnang) [inline spec/system/devises/invitations/edit_spec.rb]: don't mention any value if you don't need it. Setup factorybot.

File: spec/system/dashboard/callouts_spec.rb [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+  describe 'view all callout' do
+    it 'should list all callouts of current account' do
+      account_callout = create(:callout, account: user.account, metadata: { title: 'My Callout'})
+      other_callout = create(:callout, metadata: { title: 'Other Callout'})
+
+      visit dashboard_callouts_path
+
+      expect(page).to have_text('My Callout')
+      expect(page).not_to have_text('Other Callout')
```
Comment (samnang) [inline spec/system/dashboard/callouts_spec.rb]: this one you even can create a helper method

`expect(page).to have_callout(my_callout)`

```ruby
def have_callout(callout)
  have_selector(...) # base on css selector
end
```

File: spec/system/dashboard/callouts_spec.rb [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+    end
+
+    it 'on click new callout will open new callout page' do
+      visit 'dashboard/callouts'
+
+      click_button 'New callout'
+      expect(page).to have_current_path(new_dashboard_callout_path)
+    end
```
Comment (samnang) [inline spec/system/dashboard/callouts_spec.rb]: Better to write system/feature spec from user perspective and complete user stories.

In this case, just go to new callout page doesn't give any value, what they want from user perspective is be able to create callout.

File: spec/system/dashboard/callouts_spec.rb [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+
+      expect(page).not_to have_button('Resume')
+    end
+
+    it 'callout was already runned during session' do
+      callout = create(:callout, account: user.account, status: 'stopped')
+
+      visit dashboard_callout_path(callout)
+      callout.resume!
+      click_button 'Resume'
```
Comment (samnang) [inline spec/system/dashboard/callouts_spec.rb]: Better to not use system/feature spec level to test all edge cases and logics in the system. lower specs could help with that.

File: app/assets/javascripts/dashboard/base.js (hunk 1 above)
Comment (simleap) [reply app/assets/javascripts/dashboard/base.js]: i'm not really sure what you want me to do here.

File: app/views/dashboard/callouts/show.html.haml (hunk 8 above)
Comment (simleap) [reply app/views/dashboard/callouts/show.html.haml]: @samnang, if you have any ideas to help improve that.

File: app/views/kaminari/_gap.html.haml (hunk 9 above)
Comment (simleap) [reply app/views/kaminari/_gap.html.haml]: Yes to make it fits on mobile view.

File: app/controllers/dashboard/callouts_controller.rb [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    redirect_to dashboard_callouts_url, notice: 'Callout was successfully destroyed.'
+  end
+
+  def start
+    if @callout.start!
+      redirect_to dashboard_callout_url(@callout), notice: 'Callout was successfully started.'
+    else
+      redirect_to dashboard_callout_url(@callout), alert: 'Failed to start.'
+    end
+  end
```
Comment (samnang) [inline app/controllers/dashboard/callouts_controller.rb]: Move event actions to callout_events controller.
